subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,743 @@
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+
+	LOG_DBG("Configuration suspended, device %p", dev);
+}
+
+static void usbd_hid_resumed(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+
+	LOG_DBG("Configuration resumed, device %p", dev);
+}
+
+static void *usbd_hid_get_desc(struct usbd_class_data *const c_data,
+			       const enum usbd_speed speed)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+
+	if (speed == USBD_SPEED_HS) {
+		return ddata->hs_desc;
+	}
+
+	return ddata->fs_desc;
+}
+
+static int usbd_hid_init(struct usbd_class_data *const c_data)
+{
+	LOG_DBG("HID class %s init", c_data->name);
+
+	return 0;
+}
+
+static void usbd_hid_shutdown(struct usbd_class_data *const c_data)
+{
+	LOG_DBG("HID class %s shutdown", c_data->name);
+}
+
+static struct net_buf *hid_buf_alloc_ext(struct hid_device_data *const ddata,
+					 const uint16_t size, void *const data,
+					 const uint8_t ep)
+{
+	struct net_buf *buf = NULL;
+	struct udc_buf_info *bi;
+
+	buf = net_buf_alloc_with_data(ddata->pool_in, data, size, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}
+
+
+static struct net_buf *hid_buf_alloc(struct hid_device_data *const ddata,
+				     const uint8_t ep)
+{
+	struct net_buf *buf = NULL;
+	struct udc_buf_info *bi;
+
+	buf = net_buf_alloc(ddata->pool_out, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}
+
+static void hid_dev_output_handler(struct k_work *work)
+{
+	struct hid_device_data *ddata = CONTAINER_OF(work,
+						     struct hid_device_data,
+						     output_work);
+	struct usbd_class_data *c_data = ddata->c_data;
+	struct net_buf *buf;
+
+
+	if (!atomic_test_bit(&ddata->state, HID_DEV_CLASS_ENABLED)) {
+		return;
+	}
+
+	buf = hid_buf_alloc(ddata, hid_get_out_ep(c_data));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return;
+	}
+
+	usbd_ep_enqueue(c_data, buf);
+}
+
+static int hid_dev_submit_report(const struct device *dev,
+				 const uint16_t size, const uint8_t *const report,
+				 const bool sync)
+{
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	struct usbd_class_data *c_data = ddata->c_data;
+	struct usbd_contex *uds_ctx = usbd_class_get_ctx(c_data);
+	struct net_buf *buf;
+	int ret;
+
+	if (!atomic_test_bit(&ddata->state, HID_DEV_CLASS_ENABLED)) {
+		return -EACCES;
+	}
+
+	buf = hid_buf_alloc_ext(ddata, size, (void *)report, hid_get_in_ep(c_data));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate net_buf");
+		return -ENOMEM;
+	}
+
+	ret = usbd_ep_enqueue(c_data, buf);
+	if (ret) {
+		return usbd_ep_buf_free(uds_ctx, buf);
+	}
+
+	if (sync && ops->input_report_done == NULL) {
```

Comment:
Shouldn't it be `if (sync)` ? Currently it seems misaligned with API doc

Reply:
I updated the documentation, but kept the sync parameter for now. I am still thinking about it.

Reply:
Shouldn't it be `if (sync || ops->input_report_done == NULL)` then? (to align with newly introduced API doc)
Currently, if the callback is NULL and sync is set to false, the application has no way to get information that report was actually sent.

Reply:
I see it's already aligned. Thanks

Reply:
>Maybe we could simply use sync only if ops->input_report_done == NULL and use async otherwise (then an additional function argument may not be needed here).

I followed this suggestion.